Microsoft.Toolkit.Uwp.UI.Controls.Primitives/ConstrainedPresenter/ConstrainedPresenter.cs

```diff
@@ -0,0 +1,119 @@
+        /// </summary>
+        public AspectRatio AspectRatio
+        {
+            get { return (AspectRatio)GetValue(AspectRatioProperty); }
+            set { SetValue(AspectRatioProperty, value); }
+        }
+
+        /// <summary>
+        /// Identifies the <see cref="AspectRatio"/> property.
+        /// </summary>
+        public static readonly DependencyProperty AspectRatioProperty =
+            DependencyProperty.Register(nameof(AspectRatio), typeof(AspectRatio), typeof(ConstrainedPresenter), new PropertyMetadata(null, AspectRatioPropertyChanged));
+
+        private static void AspectRatioPropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
+        {
+            if (d is ConstrainedPresenter panel)
+            {
+                panel.InvalidateMeasure();
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the scale for the width of the panel. Should be a value between 0-1.0. Default is 1.0.
+        /// </summary>
+        public double ScaleX
+        {
+            get { return (double)GetValue(ScaleXProperty); }
+            set { SetValue(ScaleXProperty, value); }
+        }
+
+        /// <summary>
+        /// Identifies the <see cref="ScaleX"/> property.
+        /// </summary>
+        public static readonly DependencyProperty ScaleXProperty =
+            DependencyProperty.Register(nameof(ScaleX), typeof(double), typeof(ConstrainedPresenter), new PropertyMetadata(1.0, ScalePropertyChanged));
+
+        /// <summary>
+        /// Gets or sets the scale for the height of the panel. Should be a value between 0-1.0. Default is 1.0.
+        /// </summary>
+        public double ScaleY
+        {
+            get { return (double)GetValue(ScaleYProperty); }
+            set { SetValue(ScaleYProperty, value); }
+        }
+
+        /// <summary>
+        /// Identifies the <see cref="ScaleY"/> property.
+        /// </summary>
+        public static readonly DependencyProperty ScaleYProperty =
+            DependencyProperty.Register(nameof(ScaleY), typeof(double), typeof(ConstrainedPresenter), new PropertyMetadata(1.0, ScalePropertyChanged));
+
+        private static void ScalePropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
+        {
+            if (d is ConstrainedPresenter panel)
+            {
+                panel.InvalidateMeasure();
+            }
+        }
+
+        /// <inheritdoc/>
+        protected override Size MeasureOverride(Size availableSize)
+        {
+            return base.MeasureOverride(CalculateConstrainedSize(availableSize));
+        }
+
+        /// <inheritdoc/>
+        protected override Size ArrangeOverride(Size finalSize)
+        {
+            return base.ArrangeOverride(CalculateConstrainedSize(finalSize));
+        }
+
+        private Size CalculateConstrainedSize(Size initialSize)
+        {
+            var availableSize = new Size(initialSize.Width * ScaleX, initialSize.Height * ScaleY);
```

Comment:
Question: What if the content inside says "I want to be 100px wide at most"? According to docs, since we are more or less acting like a panel, we should respect that. However we are not unless base.MeasureOverride/base.ArrangeOverride does that. Thoughts?

Reply:
Pretty sure the base calls should follow those constraints, the sample I had kind of tests that scenario, but we could test some others to be sure.

Reply:
Maybe as part of some automated UI tests but should be fine then. Just a bit unsettling not knowing what exactly happens on the base calls.